Periodically check the state of regions

kitallis opened the pull request:
## Because

We don't want to get this wrong, it's crucial to proper functioning of the app.

## This addresses

Adds a layer of periodic polling to ensure everything is in good shape.

kitallis commented on config/schedule.rb:
```diff
@@ -45,3 +45,7 @@ def local(time)
   end
 end
+
+every 5.minutes, roles: [:cron] do
```
All queries on a production-sized dataset run in under 500ms. This should be safe to run on production several times.

harimohanraj89 commented on app/services/regions_integrity_check.rb:
```diff
@@ -0,0 +1,138 @@
+        (region.to_set - source.to_set).count
+      else
+        0
+      end
```
```
[
  (region.to_set - source.to_set).count,
  0
].max
```
?

harimohanraj89 commented on app/services/regions_integrity_check.rb:
```diff
@@ -0,0 +1,138 @@
+
+  def organizations
+    {
+      source: Organization.pluck(:id),
```
Should we `to_set` here at the source to simplify the inspection methods below?

harimohanraj89 commented on app/services/regions_integrity_check.rb:
```diff
@@ -0,0 +1,138 @@
+      {
+        missing_regions: [],
+        missing_sources_count: 0
+      }
```
I'd just shove this data structure into `initialize`. Don't think we need the indirection.

`ok?` can ensure the first key is empty and the second key is zero.

What do you think?

harimohanraj89 commented on app/services/regions_integrity_check.rb:
```diff
@@ -0,0 +1,138 @@
+
+    def set_inconsistencies
+      @inconsistencies[:missing_regions] += sources_without_regions
+      @inconsistencies[:missing_sources_count] = regions_without_sources_count
```
Would this be more helpful if it collected a list of region IDs we can inspect, rather than a count?

harimohanraj89 commented on app/services/regions_integrity_check.rb:
```diff
@@ -0,0 +1,138 @@
+
+  def facilities
+    {
+      source: Facility.pluck(:id),
```
How do we need to use `with_discarded` for these integrity checks?

kitallis replied:
Nah, I think neither `RegionBackfill` nor callbacks care about discarded records, so it wouldn't matter if they went out of sync anyway. I **think** it's ok to not have it.

harimohanraj89 replied:
Does deleting facilities delete the region too? I'm thinking if we delete a facility, will we get a "region without source" show up in this check?

kitallis replied:
[RegionSource](https://github.com/simpledotorg/simple-server/blob/master/app/models/region_source.rb#L4) handles this.

harimohanraj89 reviewed (changes requested):
Drats, left standalone comments by mistake. 

Review above! ☝️

kitallis commented:
> Drats, left standalone comments by mistake.
> 
> Review above! ☝️

@harimohanraj89 So I changed the script drastically because it wasn't letting me mold it freely. I rewrote parts of it and removed the test that checks for "missing sources"; it instead checks for "duplicate regions" which is more valuable. Regions without sources is harmless, so not really something getting scared about.

I also benched the queries overall on prod-sized data and this was my samples:

```
# benchmark { RegionsIntegrityCheck.sweep }

I, [2020-12-08T14:48:08.348171 #45463]  INFO -- : Benchmarking (1012.2ms)
I, [2020-12-08T14:49:38.327484 #45463]  INFO -- : Benchmarking (1106.5ms)
I, [2020-12-08T14:49:45.556005 #45463]  INFO -- : Benchmarking (967.6ms)
I, [2020-12-08T14:49:51.610416 #45463]  INFO -- : Benchmarking (1015.1ms)
I, [2020-12-08T14:50:00.146900 #45463]  INFO -- : Benchmarking (989.6ms)
I, [2020-12-08T14:50:06.171613 #45463]  INFO -- : Benchmarking (1071.7ms)
I, [2020-12-08T14:50:20.987763 #45463]  INFO -- : Benchmarking (1056.3ms)
I, [2020-12-08T14:50:28.620415 #45463]  INFO -- : Benchmarking (1105.1ms)
```

Since it has increased in run time by 500ms from before, I have changed the schedule from 5 minutes to every hour.

harimohanraj89 commented on app/services/regions_integrity_check.rb:
```diff
@@ -0,0 +1,198 @@
+    missing_regions =
+      (Organization.pluck(:id).to_set -
+        Region.organization_regions.pluck(:source_id).to_set
+      ).to_a
```
Do you need to `to_set`? Does that handle duplicates?

harimohanraj89 reviewed (approved):
Walked through this sanity check with you on zoom. Looks cooooool.
